### ISSUE TITLE
RGB contrast is not well managed

### ISSUE THREAD

**Issue by tanguyduval** [napari/napari#1545]
## 🐛 Bug
Intensity range is not well managed for RGB images encoded as:

- [ ] uint16
- [ ] float32

For example on a 16bits images, with values around 1000, contrast need to be set to [0-128] to see something, and colors are flattened...:
![image](https://user-images.githubusercontent.com/7785316/90109995-dd672980-dd4c-11ea-97a1-14f377e2525d.png)
While on an other viewer the image look like this with an intensity range of [0 2250]
![image](https://user-images.githubusercontent.com/7785316/90110070-f7087100-dd4c-11ea-9c7b-a304cb19a28f.png)
 
## To Reproduce

1. write a 16bit image: `imageio.imwrite('result.tif',np.uint16(data.astronaut())*10)`
or a float 32 image: `imageio.imwrite('~/result.tif',np.float32(data.astronaut())/3)`
2. open the 16bits rgb image through drag and drop or `add_image`:
````
import numpy as np
from skimage import data
viewer.add_image(np.float32(data.astronaut())/3,rgb=True) # RGB
viewer.add_image(np.transpose(np.float32(data.astronaut())/3,[2, 0, 1]),rgb=False) # GRAY FOR COMPARISON
````
3. reduce contrast upper limit until you see the image. Image quality is poor and pixel value is saturated to 1 for float32 image... 

## Expected behavior

Correctly display colors

**Comment by trebconnell** [comment]
I believe I hit the same issue.  But have a minimal repro that may be useful when resolving.  If the image is viewed initially with constrained contrast_limits it visualizes nicely.  If the contrast_limits are changed in the UI then results are not ideal.

To reproduce run following code with napari 0.4.7 and change contrast limits to 0-150 in UI.  As shown in image below.:
```
import numpy
data = numpy.array([[0, 1, 2, 3, 4, 5, 10, 100, 10000]], numpy.uint16)
data = numpy.repeat(numpy.repeat(data, 20, axis=0), 20, axis=1)
import napari
# Changing contrast limits to 0-150 in UI doesn't work
viewer = napari.view_image(data, contrast_limits=[0,10000], colormap='turbo')
# This works
#viewer = napari.view_image(data, contrast_limits=[0,150], colormap='turbo')
napari.run()
```

Result via changing limits to 0-150 in UI:
![image](https://user-images.githubusercontent.com/6809385/118176014-b3c75580-b3e5-11eb-92b2-7b586c501563.png)
Result via changing limits to 0-150 in view_image:
![image](https://user-images.githubusercontent.com/6809385/118176111-d78a9b80-b3e5-11eb-96e7-5a41105d43f5.png)

**Comment by trebconnell** [comment]
I found a workaround.  Can apply Reset View (ctrl+r) after setting the contrast limits in the UI.

**Comment by tlambert03** [comment]
ah interesting... this is an issue with the vispy rendering with big changes in contrast limits.  
It's not actually directly connected to whether the clims are changed in the UI or via code... it has more to do with hitting that clims of (0, 10000), and the image texture not "coming back".  Will need to dig in more later, but here's an example of "rescuing" that first example:

```python
import numpy
import napari

data = numpy.array([[0, 1, 2, 3, 4, 5, 10, 100, 10000]], numpy.uint16)
data = numpy.repeat(numpy.repeat(data, 20, axis=0), 20, axis=1)

viewer = napari.view_image(data, contrast_limits=[0,10000], colormap='turbo')
layer.contrast_limits = (0, 50) # don' work ... but notice that the thumbnail looks fine

# rebuild the GPU texture:
visual = viewer.window.qt_viewer.layer_to_visual[layer]
visual.node._build_texture()
visual.node.update()   # now the image looks like the thumbail
```

thanks for the reproducible example!  will be able to fix soonish

**Comment by tanguyduval** [comment]
Thank you for digging this issue out! It seems like a different issue to me. I tried ctrl+R or rebuilding the texture and it did not work (napari 0.4.7).

I realized that my examples were not very clean.  
  
Here how to reproduce the issue for uint16 images:
1. open napari
2. in the napari ipython terminal:
````python
import numpy as np
from skimage import data
viewer.add_image(np.uint16(data.astronaut()),rgb=True) # RGB
````
 image is black...
  
I found a different but similar issue with float images:
1. open napari
2. in the napari ipython terminal:
````python
import numpy as np
from skimage import data
viewer.add_image(np.float32(data.astronaut()),rgb=True) # RGB
````

values are saturated to 1... I know some other viewers can have this behavior but this is not what you want when using physical images.

Best
